### PR TITLE
fix(NODE-6355): respect utf8 validation options when iterating cursors

### PR DESCRIPTION
**Pull request by baileympearson**
#### What is changing?

This PR ensures that any BSON utf8 validation options are provided to the deserializer when deserializing cursors.

Since on-demand BSON, cursors are deserialized on-demand.  Documents are pulled out of the cursor using `MongoDBResponse.get()`, which returns an unparsed on-demand document.  We then deserialize it using `toObject()`.

However, although `MongoDBResponse` correctly applies BSON utf8 options to its deserialization logic, the on-demand document does not.  This results in documents retrieved from a cursor being deserialized without the utf8 validation options.

### Double check the following

- [ ] Ran `npm run check:lint` script
- [ ] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [ ] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/integration/node-specific/bson-options/utf8_validation.test.ts [hunk 1]
```diff
@@ -153,3 +156,177 @@ describe('class MongoDBResponse', () => {
...
+   * bytes of the character 'é', to produce invalid utf8.
+   */
+  async function insertDocumentWithInvalidUTF8() {
+    const stub = sinon.stub(net.Socket.prototype, 'write').callsFake(function (...args) {
+      const providedBuffer = args[0].toString('hex');
+      const targetBytes = Buffer.from('é').toString('hex');
+
+      if (providedBuffer.includes(targetBytes)) {
+        if (providedBuffer.split(targetBytes).length !== 2) {
+          throw new Error('received buffer more than one `c3a9` sequences.  or perhaps none?');
```
Comment (nbbeeken) [inline test/integration/node-specific/bson-options/utf8_validation.test.ts:176]:
```suggestion
          sinon.restore();
          throw new Error('received buffer more than one `c3a9` sequences. or perhaps none?');
```

Maybe this is why we didn't see the error?

File: test/integration/node-specific/bson-options/utf8_validation.test.ts [hunk 2]
```diff
@@ -153,3 +156,177 @@ describe('class MongoDBResponse', () => {
...
+    await client.connect();
+    const db = client.db('test');
+    collection = db.collection('invalidutf');
+
+    await collection.deleteMany({});
+    await insertDocumentWithInvalidUTF8();
+  });
+
+  afterEach(async function () {
+    await client.close();
```
Comment (nbbeeken) [inline test/integration/node-specific/bson-options/utf8_validation.test.ts:207]:
```suggestion
    sinon.restore();
    await client.close();
```

Just to be extra sure we revert to normal.

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: test/integration/node-specific/bson-options/utf8_validation.test.ts (hunk 1 above)
Comment (baileympearson) [reply test/integration/node-specific/bson-options/utf8_validation.test.ts:176]: Once this method returns / errors, we'll still have to continue up the sinon callstack back into the driver, regardless of whether or not the stub has been restored.

**Review by nbbeeken — CHANGES_REQUESTED** [review]
The `c3a9` condition is failing. Perhaps we need to be a bit more robust, searching for `c3a9` after finding `"insert"` or `"documents"` so we know we're in the command.

**Comment by jia-zhen** [comment]
Please update and release it as soon as possible, as we are currently troubled by this issue. / Please update and publish it promptly, as we are being bothered by this problem.

File: .evergreen/install-mongodb-client-encryption.sh [hunk 3]
```diff
@@ -10,7 +10,7 @@ set -o xtrace   # Write all commands first to stderr
 set -o errexit  # Exit the script with error if any of the commands fail
 
 rm -rf mongodb-client-encryption
-git clone https://github.com/mongodb-js/mongodb-client-encryption.git
+git clone https://github.com/mongodb-js/mongodb-client-encryption.git -b explicit-lifetime-chaining
```
Comment (nbbeeken) [inline .evergreen/install-mongodb-client-encryption.sh:13]: Seems unrelated, accidentally added this when testing the fle PR?

File: src/cmap/wire_protocol/on_demand/document.ts [hunk 4]
```diff
@@ -45,6 +46,13 @@ type CachedBSONElement = { element: BSONElement; value: any | undefined };
 
 /** @internal */
 export class OnDemandDocument {
+  /**
+   * @internal
+   *
+   * Used for testing purposes.
+   */
+  private static BSON: typeof BSON = BSON;
```
Comment (nbbeeken) [inline src/cmap/wire_protocol/on_demand/document.ts:54]: Why do we need to do this but not something similar for `net`?

Comment (nbbeeken) [reply src/cmap/wire_protocol/on_demand/document.ts:54]: Let's determine the reason before introducing this injectable. Are you able to debug? I will take a look when I'm at my desk

**Review by nbbeeken — APPROVED** [review]
LGTM
